apps/google-analytics-4/frontend/src/components/config-screen/HomeAnalyticsPage.tsx

```diff
@@ -0,0 +1,218 @@
+    };
+
+    const newParameters = Object.assign(
+      {},
+      parameters,
+      omitBy(newServiceKeyParameters, (val) => val === null)
+    );
+
+    setIsInEditMode(false);
+    setParameters(newParameters);
+    setServiceAccountKeyFileIsRequired(false);
+    setServiceAccountKeyFile('');
+
+    return {
+      parameters: newParameters,
+      targetState: currentState,
+    };
+  }, [
+    newServiceAccountKey,
+    newServiceAccountKeyId,
+    serviceAccountKeyFileIsRequired,
+    serviceAccountKeyFileIsValid,
+    parameters,
+    sdk,
+  ]);
+
+  useEffect(() => {
+    sdk.app.onConfigure(() => onConfigure());
+  }, [sdk, onConfigure]);
+
+  useEffect(() => {
+    const setupAppInstallationParameters = async () => {
+      const currentParameters: AppInstallationParameters | null = await sdk.app.getParameters();
+
+      if (currentParameters) {
+        setParameters(currentParameters);
+        setServiceAccountKeyFileIsRequired(false);
+      } else {
+        // per the documentation, `null` means app is not installed, thus we will require
+        // the key file
+        setServiceAccountKeyFileIsRequired(true);
+      }
+
+      sdk.app.setReady();
+    }
+
+    setupAppInstallationParameters();
+  }, [sdk]);
+
+  const setValidServiceAccountKey = (newServiceAccountKey: ServiceAccountKey | null) => {
+    setNewServiceAccountKey(newServiceAccountKey);
+    setNewServiceAccountKeyId(
+      newServiceAccountKey
+        ? convertServiceAccountKeyToServiceAccountKeyId(newServiceAccountKey)
+        : null
+    );
+    setServiceAccountKeyFileErrorMessage('');
+    setServiceAccountKeyFileIsValid(true);
+  };
+
+  const setInvalidServiceAccountKey = (errorMessage: string) => {
+    setNewServiceAccountKey(null);
+    setNewServiceAccountKeyId(null);
+    setServiceAccountKeyFileErrorMessage(errorMessage);
+    setServiceAccountKeyFileIsValid(false);
+  };
+
+  const handleKeyFileChangeEventWrapper = (event: React.ChangeEvent<HTMLInputElement>) => {
+    handleKeyFileChange(event.target.value);
+  };
+
+  const handleKeyFileChange = (keyFile: string) => {
+    setServiceAccountKeyFile(keyFile);
+
+    const trimmedFieldValue = keyFile;
+    if (trimmedFieldValue === '') {
+      setValidServiceAccountKey(null);
+      return;
+    }
+
+    let newServiceAccountKey: ServiceAccountKey;
+    try {
+      newServiceAccountKey = convertKeyFileToServiceAccountKey(trimmedFieldValue);
+    } catch (e) {
+      if (
+        // failed assertions about key file contents
+        e instanceof AssertionError ||
+        // could not parse as JSON
+        e instanceof SyntaxError
+      ) {
+        setInvalidServiceAccountKey(e.message);
+      } else {
+        console.error(e);
+        setInvalidServiceAccountKey('An unknown error occurred');
+      }
+      return;
+    }
+
+    setValidServiceAccountKey(newServiceAccountKey);
+  };
+
+  const handleEditGoogleAccountDetails = () => {
+    setIsInEditMode(true);
```

Comment:
Common Pattern Comment: Function definitions use the `handle` keyword and then the `on` keyword when passed as a prop/callback. 

Ctrl+F `handleEditGoogleAccountDetails` and `onEditGoogleAccountDetails` to see an example

Reply:
you mean `CMD + F`? ;)